src/uct/ib/dc/dc_mlx5_ep.h

```diff
@@ -300,86 +300,92 @@ uct_dc_mlx5_ep_from_dci(uct_dc_mlx5_iface_t *iface, uint8_t dci_index)
 }
 
 void uct_dc_mlx5_ep_handle_failure(uct_dc_mlx5_ep_t *ep,
                                    struct mlx5_cqe64 *cqe,
                                    ucs_status_t status);
 
 static UCS_F_ALWAYS_INLINE void
-uct_dc_mlx5_init_dci_config_key(uct_dc_mlx5_dci_config_t *dci_config,
-                                uint8_t path_index,
-                                uint8_t is_keepalive,
-                                uint8_t max_rd_atomic)
+uct_dc_mlx5_init_dci_config(uct_dc_mlx5_dci_config_t *dci_config,
+                            uint8_t path_index, uint8_t max_rd_atomic)
 {
-    dci_config->key.path_index = path_index;
-    dci_config->key.flags      = 0;
-
-    if (is_keepalive) {
-        dci_config->key.flags |= UCT_DC_MLX5_DCI_CONFIG_KEEPALIVE;
-    }
-
-    if (max_rd_atomic == 64) {
-        dci_config->key.flags |= UCT_DC_MLX5_DCI_CONFIG_MAX_RD_ATOMIC_IS_64;
-    }
-
-    memset(dci_config->key.padding, 0, sizeof(dci_config->key.padding));
+    dci_config->path_index    = path_index;
+    dci_config->max_rd_atomic = max_rd_atomic;
 }
 
 ucs_status_t static UCS_F_ALWAYS_INLINE
 uct_dc_mlx5_dci_pool_add_dci(uct_dc_mlx5_iface_t *iface, uint8_t pool_index,
                              uint8_t dci_index)
 {
     uct_dc_mlx5_dci_pool_t *pool = &iface->tx.dci_pool[pool_index];
-    uct_dc_mlx5_dci_config_t pool_config = {
-        .u64 = pool->config_key
-    };
+    uct_dc_dci_t *dci            = uct_dc_mlx5_iface_dci(iface, dci_index);
     ucs_status_t status;
-    uint8_t* stack_ptr;
 
-    ucs_assertv(ucs_array_length(&pool->stack) <= iface->tx.ndci,
+    ucs_assertv(ucs_array_length(&pool->stack) < iface->tx.ndci,
                 "stack length exceeded ndci");
 
-    if ((ucs_array_capacity(&pool->stack) == ucs_array_length(&pool->stack)) ||
-        uct_dc_mlx5_is_dci_valid(&ucs_array_elem(&iface->tx.dcis, dci_index))) {
-        return UCS_OK;
-    }
-
-    status = uct_dc_mlx5_iface_create_dci(
-            iface, pool_index, dci_index,
-            pool_config.key.path_index,
-            iface->flags & UCT_DC_MLX5_IFACE_FLAG_DCI_FULL_HANDSHAKE);
-
+    status = uct_dc_mlx5_iface_create_dci(iface, dci_index, dci);
     if (status != UCS_OK) {
         ucs_error("iface %p: failed to create dci %u at pool %u", iface,
                   dci_index, pool_index);
         return status;
     }
+    dci->path_index = pool->config.path_index;
+    dci->pool_index = pool_index;
 
-    stack_ptr  = ucs_array_append(&pool->stack, return UCS_ERR_NO_MEMORY);
-    *stack_ptr = dci_index;
+    if(!uct_dc_mlx5_iface_is_dci_rand(iface)) {
+        *ucs_array_append(&pool->stack, return UCS_ERR_NO_MEMORY) = dci_index;
+    }
 
     return UCS_OK;
 }
 
+ucs_status_t static UCS_F_ALWAYS_INLINE
+uct_dc_mlx5_dci_pool_append_dci(uct_dc_mlx5_iface_t *iface, uint8_t pool_index)
+{
+    uct_dc_dci_t dci  = {{{0}}};
+    uint8_t dci_index = ucs_array_length(&iface->tx.dcis);
+
+    dci.txwq.super.qp_num = UCT_IB_INVALID_QPN;
```

Comment:
seems like the logic of adding empty dci appears 3 times in the code, can we unite?